

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Example using scidat: Download Annotate TCGA &mdash; scidat 1.0.2 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script src="../_static/custom.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <!-- Global site tag (gtag.js) - Google Analytics -->
<!--    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-84847737-2"></script>-->
<!--    <script>-->
<!--        window.dataLayer = window.dataLayer || [];-->

<!--        function gtag() {-->
<!--            dataLayer.push(arguments);-->
<!--        }-->

<!--        gtag('js', new Date());-->

<!--        gtag('config', 'UA-84847737-2');-->
<!--    </script>-->


    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="FAQ" href="../faq.html" />
    <link rel="prev" title="CLI" href="cli.html" />
 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/logo.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.0.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../about.html">sci-dat</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installing/index.html">Installing sci-DAT</a></li>
</ul>
<p class="caption"><span class="caption-text">Running sci-dat</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="examples.html">R Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">CLI</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Example using scidat: Download Annotate TCGA</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Section-1:-Data-selection-and-metadata-download">Section 1: Data selection and metadata download</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Download:">Download:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Get-GDC-Data-Transfer-Tool">Get GDC Data Transfer Tool</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Section-2:-File-download">Section 2: File download</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Section-3:-Clinical-information-annotation">Section 3: Clinical information annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Section-4:-Annotation-with-mutation-data">Section 4: Annotation with mutation data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Look-at-the-data-and-add-gene-names">Look at the data and add gene names</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Example-1:-Calculating-the-logfold-change-accross-2-conditions">Example 1: Calculating the logfold change accross 2 conditions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Example-2:-Looking-at-differences-across-genes-between-conditions">Example 2: Looking at differences across genes between conditions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Example-3:-Looking-at-genes-across-conditions">Example 3: Looking at genes across conditions</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">scidat</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content style-external-links">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Example using scidat: Download Annotate TCGA</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/arianemora/scidat/blob/master/docs_src/src/examples/notebook.ipynb" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Example-using-scidat:-Download-Annotate-TCGA">
<h1>Example using scidat: Download Annotate TCGA<a class="headerlink" href="#Example-using-scidat:-Download-Annotate-TCGA" title="Permalink to this headline">¶</a></h1>
<p>This was designed to be a simple interface that allows for downloading and annotating data from TCGA.</p>
<div class="section" id="Section-1:-Data-selection-and-metadata-download">
<h2>Section 1: Data selection and metadata download<a class="headerlink" href="#Section-1:-Data-selection-and-metadata-download" title="Permalink to this headline">¶</a></h2>
<p>Download metatdata from TCGA. For this program to work you need to first go to: <a class="reference external" href="https://portal.gdc.cancer.gov/">https://portal.gdc.cancer.gov/</a> and select the files to download.</p>
<p>In this example, I chose my data by using the following steps: 1. Navigate to <code class="docutils literal notranslate"><span class="pre">Exploration</span></code> tab 2. Selected <code class="docutils literal notranslate"><span class="pre">kidney</span></code> as the Primary Site 3. Selected <code class="docutils literal notranslate"><span class="pre">solid</span> <span class="pre">tissue</span> <span class="pre">normal</span></code> and <code class="docutils literal notranslate"><span class="pre">primary</span> <span class="pre">tumor</span></code> for my Sample Type (823 cases) 4. Selected <code class="docutils literal notranslate"><span class="pre">KIRP</span></code> as the project (279 cases) 5. Selected <code class="docutils literal notranslate"><span class="pre">dead</span></code> as the vital status (43 cases) 6. Clicked <code class="docutils literal notranslate"><span class="pre">View</span> <span class="pre">Files</span> <span class="pre">in</span> <span class="pre">Repository</span></code></p>
<div class="line-block">
<div class="line">Since I’m only interested in RNAseq count data and Methylation beta values, I filter my files to only include these two types of files.</div>
<div class="line">7. Selected <code class="docutils literal notranslate"><span class="pre">RNA-seq</span></code> as my Experimental Strategy 8. Selected <code class="docutils literal notranslate"><span class="pre">HTSeq</span> <span class="pre">-</span> <span class="pre">Counts</span></code> in Workflow Type 9. Clicked <code class="docutils literal notranslate"><span class="pre">Add</span> <span class="pre">all</span> <span class="pre">files</span> <span class="pre">to</span> <span class="pre">cart</span></code> (50 files, 43 cases)</div>
</div>
<p>link:
<a class="reference external" href="https://portal.gdc.cancer.gov/repository?filters=%7B%22op%22%3A%22and%22%2C%22content%22%3A%5B%7B%22content%22%3A%7B%22field%22%3A%22cases.case_id%22%2C%22value%22%3A%5B%22set_id%3AAXGBi619om_VEIzbp9eW%22%5D%7D%2C%22op%22%3A%22IN%22%7D%2C%7B%22op%22%3A%22in%22%2C%22content%22%3A%7B%22field%22%3A%22files.analysis.workflow_type%22%2C%22value%22%3A%5B%22HTSeq%20-%20Counts%22%5D%7D%7D%2C%7B%22op%22%3A%22in%22%2C%22content%22%3A%7B%22field%22%3A%22files.experimental_strategy%22%2C%22value%22%3A%5B%22RNA-Seq%22%5D%7D%7D%5D%7D">https://portal.gdc.cancer.gov/repository?filters=%7B%22op%22%3A%22and%22%2C%22content%22%3A%5B%7B%22content%22%3A%7B%22field%22%3A%22cases.case_id%22%2C%22value%22%3A%5B%22set_id%3AAXGBi619om_VEIzbp9eW%22%5D%7D%2C%22op%22%3A%22IN%22%7D%2C%7B%22op%22%3A%22in%22%2C%22content%22%3A%7B%22field%22%3A%22files.analysis.workflow_type%22%2C%22value%22%3A%5B%22HTSeq%20-%20Counts%22%5D%7D%7D%2C%7B%22op%22%3A%22in%22%2C%22content%22%3A%7B%22field%22%3A%22files.experimental_strategy%22%2C%22value%22%3A%5B%22RNA-Seq%22%5D%7D%7D%5D%7D</a></p>
<p>Now I navigated to my cart which had 50 files in it. Since it is recomended to use th GDC-data transfer tool (which scidat uses) I only download the <code class="docutils literal notranslate"><span class="pre">manifest</span></code> and the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> for my files of interest. Before proceeding, look at how much space the files will need (e.g. for me this is 12.62MB) make sure the computer you are downloading on has that space available.</p>
<div class="section" id="Download:">
<h3>Download:<a class="headerlink" href="#Download:" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>13. `Biospecimen` (tsv)
14. `Clinical` (tsv)
15. `Sample Sheet`
16. `Metadata`
17. Click the `Download` button and select `Manifest`
</pre></div>
</div>
<p>Last steps: 1. Lastly, move all these files to a new empty directory <code class="docutils literal notranslate"><span class="pre">~/Documents/TCGA_data_download_scidat/</span></code>. 2. Unzip the <code class="docutils literal notranslate"><span class="pre">Clinical</span></code> folder and delete the zipped version. 3. Copy the <code class="docutils literal notranslate"><span class="pre">clinical.tsv</span></code> file in the <code class="docutils literal notranslate"><span class="pre">clinical.cart....</span></code> folder to the <code class="docutils literal notranslate"><span class="pre">~/Documents/TCGA_data_download_scidat/</span></code> directory. 4. Make a new directory in <code class="docutils literal notranslate"><span class="pre">~/Documents/TCGA_data_download_scidat/</span></code> called <code class="docutils literal notranslate"><span class="pre">downloads</span></code> (we’ll use this below)</p>
</div>
</div>
<div class="section" id="Get-GDC-Data-Transfer-Tool">
<h2>Get GDC Data Transfer Tool<a class="headerlink" href="#Get-GDC-Data-Transfer-Tool" title="Permalink to this headline">¶</a></h2>
<p>If you haven’t already got the GDC transfer tool, you’ll need to download this, follow the instructions from TCGA on how to do this: <a class="reference external" href="https://gdc.cancer.gov/access-data/gdc-data-transfer-tool">https://gdc.cancer.gov/access-data/gdc-data-transfer-tool</a></p>
<p>(choose the correct version of the below 3 for your system) Package icongdc-client_v1.5.0_Windows_x64.zip Package icongdc-client_v1.5.0_Ubuntu_x64.zip Package icongdc-client_v1.5.0_OSX_x64.zip</p>
<p>Unzip and move the downloaded GDC transfer tool to the same directory you put your manifest file in i.e. <code class="docutils literal notranslate"><span class="pre">~/Documents/TCGA_data_download_scidat/</span></code> from before. Now we’re ready to download the files and annotate them with the data we downloaded.</p>
</div>
<div class="section" id="Section-2:-File-download">
<h2>Section 2: File download<a class="headerlink" href="#Section-2:-File-download" title="Permalink to this headline">¶</a></h2>
<p>The script below assumes you have put the files from above in <code class="docutils literal notranslate"><span class="pre">~/Documents/TCGA_data_download_scidat/</span></code> and that your <code class="docutils literal notranslate"><span class="pre">gdc-client</span></code> is executable.</p>
<p>If you are using a windows machine (or put your files in a different folder you’ll need to rename that directory).</p>
<p>Before running the script below you will need to make sure you have edited the path to the manifest file, and the gdc_client to be located where you placed them on your computer and with their proper names (e.g.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">scidat</span> <span class="kn">import</span> <span class="n">API</span>

<span class="c1"># SET THIS TO TRUE FOR THE FIRST TIME YOU RUN THIS</span>
<span class="n">download</span> <span class="o">=</span> <span class="kc">False</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This will take about 3 minutes so sit back and have a coffee!</span>

<span class="sd">Note if you have a very large amount of data to download you should increase the max_cnt.</span>
<span class="sd">This will run a process for each 5 cases, i.e. if your downloading 5000 files, that will be 1000 processes!</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="n">meta_dir</span> <span class="o">=</span> <span class="s1">&#39;/Users/ariane/Documents/code/tcga-format/data/example_download/TCGA_data_download_scidat_unix/&#39;</span>
<span class="n">manifest_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">meta_dir</span><span class="si">}</span><span class="s1">gdc_manifest_20200416_055430.txt&#39;</span>
<span class="n">gdc_client</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">meta_dir</span><span class="si">}</span><span class="s1">./gdc-client&#39;</span>
<span class="n">download_dir</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">meta_dir</span><span class="si">}</span><span class="s1">downloads/&#39;</span>

<span class="n">clinical_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">meta_dir</span><span class="si">}</span><span class="s1">clinical.tsv&#39;</span>
<span class="n">sample_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">meta_dir</span><span class="si">}</span><span class="s1">gdc_sample_sheet.2020-04-16.tsv&#39;</span>
<span class="n">annotation_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">meta_dir</span><span class="si">}</span><span class="s1">tcga_hsapiens_gene_ensembl-GRCh38.p13.csv&#39;</span>

<span class="c1"># scidat spits the manifest file into submanifests so that we can make multiple calls to TCGA simulateously.</span>
<span class="c1"># This speeds up the download process. Note, it will use up more of your computers&#39; processing so if you are worried</span>
<span class="c1"># just set the max_cnt to be more than the number of files in your manifest e.g. 100000000</span>
<span class="n">api</span> <span class="o">=</span> <span class="n">API</span><span class="p">(</span><span class="n">manifest_file</span><span class="p">,</span> <span class="n">gdc_client</span><span class="p">,</span> <span class="n">clinical_file</span><span class="p">,</span> <span class="n">sample_file</span><span class="p">,</span>  <span class="n">download_dir</span><span class="o">=</span><span class="n">download_dir</span><span class="p">,</span> <span class="n">annotation_file</span><span class="o">=</span><span class="n">annotation_file</span><span class="p">,</span> <span class="n">requires_lst</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">clin_cols</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
          <span class="n">max_cnt</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">sciutil</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">split_manifest_dir</span><span class="o">=</span><span class="n">download_dir</span><span class="p">,</span>
          <span class="n">meta_dir</span><span class="o">=</span><span class="n">meta_dir</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>

<span class="k">if</span> <span class="n">download</span><span class="p">:</span>
    <span class="n">api</span><span class="o">.</span><span class="n">download_data_from_manifest</span><span class="p">()</span>

</pre></div>
</div>
</div>
</div>
<div class="section" id="Section-3:-Clinical-information-annotation">
<h2>Section 3: Clinical information annotation<a class="headerlink" href="#Section-3:-Clinical-information-annotation" title="Permalink to this headline">¶</a></h2>
<p>Here we are going to builld the clinical information dataframe.</p>
<p>We’re also interested in the mutation data so we’re going to download the mutation data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>

<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This section will take about 30 second to 1 min</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="c1"># Build the annotation</span>
<span class="n">api</span><span class="o">.</span><span class="n">build_annotation</span><span class="p">()</span>

<span class="c1"># Download the mutation data --&gt; here we can pass only some case IDs or if we leave it empty we download all the</span>
<span class="c1"># mutation data for our cases.</span>
<span class="k">if</span> <span class="n">download</span><span class="p">:</span>
    <span class="n">api</span><span class="o">.</span><span class="n">download_mutation_data</span><span class="p">()</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-blue-intense-fg">--------------------------------------------------------------------------------</span>
<span class="ansi-blue-intense-fg">                              Clinical dataframe                                    </span>
<span class="ansi-blue-intense-fg">--------------------------------------------------------------------------------</span>
<span class="ansi-blue-intense-fg">--------------------------------------------------------------------------------</span>
<span class="ansi-blue-intense-fg">   submitter_id project_id  age_at_index  gender                       race  \
0  TCGA-G7-6793  TCGA-KIRP            60  female                      white
2  TCGA-A4-7585  TCGA-KIRP            67    male                      white
4  TCGA-AL-3466  TCGA-KIRP            41    male                      white
6  TCGA-J7-8537  TCGA-KIRP            37  female                      asian
8  TCGA-BQ-5877  TCGA-KIRP            60    male  black or african american

  vital_status tumor_stage  days_to_death
0         Dead    stage iv            336
2         Dead   stage iii           1070
4         Dead    stage iv            293
6         Dead   stage iii            321
8         Dead    stage iv            270       </span>
<span class="ansi-blue-intense-fg">--------------------------------------------------------------------------------</span>
<span class="ansi-blue-intense-fg">--------------------------------------------------------------------------------</span>
<span class="ansi-blue-intense-fg">                               Sample dataframe                                      </span>
<span class="ansi-blue-intense-fg">--------------------------------------------------------------------------------</span>
<span class="ansi-blue-intense-fg">--------------------------------------------------------------------------------</span>
<span class="ansi-blue-intense-fg">                                File ID  \
0  661829b5-a067-467f-aa58-518c8f91bad9
1  0aa7c232-98e9-411e-a5a3-77dc841ce743
2  8a8c6c6c-971a-4045-b04d-9325195a1598
3  b6a0c9ec-3294-4a27-84ca-b59183c3a737
4  cf683af3-6dd7-4f28-be42-cbc1cecc560c

                                           File Name            Data Category  \
0  61e5178d-86bc-4229-8a41-61401f71f64b.htseq.cou...  Transcriptome Profiling
1  a05ad788-3f8e-4b01-9ddd-fd6643efc4c7.htseq.cou...  Transcriptome Profiling
2  7a3c64e7-98bc-4cbf-8a2a-0c3314d46ad2.htseq.cou...  Transcriptome Profiling
3  66f56582-eab0-4192-9c2a-0029c343b815.htseq.cou...  Transcriptome Profiling
4  acf91e5b-fe69-4064-9a7f-eb4094c7ee62.htseq.cou...  Transcriptome Profiling

                        Data Type Project ID       Case ID         Sample ID  \
0  Gene Expression Quantification  TCGA-KIRP  TCGA-BQ-5877  TCGA-BQ-5877-01A
1  Gene Expression Quantification  TCGA-KIRP  TCGA-5P-A9JY  TCGA-5P-A9JY-01A
2  Gene Expression Quantification  TCGA-KIRP  TCGA-BQ-5879  TCGA-BQ-5879-11A
3  Gene Expression Quantification  TCGA-KIRP  TCGA-A4-7585  TCGA-A4-7585-01A
4  Gene Expression Quantification  TCGA-KIRP  TCGA-UZ-A9PN  TCGA-UZ-A9PN-01A

           Sample Type
0        Primary Tumor
1        Primary Tumor
2  Solid Tissue Normal
3        Primary Tumor
4        Primary Tumor          </span>
<span class="ansi-blue-intense-fg">--------------------------------------------------------------------------------</span>
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Now we can do all sorts of things for example get cases with specific metadata</span>
<span class="c1"># Now we want to check the functionality we expect</span>
<span class="n">metadata</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;race&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;white&#39;</span><span class="p">],</span> <span class="s1">&#39;gender&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;male&#39;</span><span class="p">],</span> <span class="s1">&#39;project_id&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;TCGA-KIRP&#39;</span><span class="p">]}</span>

<span class="n">case_files</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_files_with_meta</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span> <span class="s2">&quot;all&quot;</span><span class="p">)</span>
<span class="n">case_files</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">case_files</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
26
</pre></div></div>
</div>
</div>
<div class="section" id="Section-4:-Annotation-with-mutation-data">
<h2>Section 4: Annotation with mutation data<a class="headerlink" href="#Section-4:-Annotation-with-mutation-data" title="Permalink to this headline">¶</a></h2>
<p>Unfortunately TCGA doesn’t make it easy to add the mutation annotation data.</p>
<p>If in <strong>Section 1</strong> you didn’t do the optional <em>step 5</em> please go back and download the mutation data as we need this to add the mutation information.</p>
<p>How this section works is that we have the different mutations (the file we downloaded from <strong>Section 1</strong> has the mutation ID and the information about the specific mutation. Unfortunatly, this isn’t connected to a case (patient). In order to connect the mutation with the patient we need to call the API provided by TCGA (<a class="reference external" href="https://docs.gdc.cancer.gov/API/Users_Guide/Data_Analysis/#simple-somatic-mutation-endpoint-examples">https://docs.gdc.cancer.gov/API/Users_Guide/Data_Analysis/#simple-somatic-mutation-endpoint-examples</a>).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># First lets build the mutation DF</span>
<span class="n">api</span><span class="o">.</span><span class="n">build_mutation_df</span><span class="p">()</span>

<span class="c1"># Lets get all the genes with mutations</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_genes_with_mutations</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
3090
</pre></div></div>
</div>
</div>
<div class="section" id="Look-at-the-data-and-add-gene-names">
<h2>Look at the data and add gene names<a class="headerlink" href="#Look-at-the-data-and-add-gene-names" title="Permalink to this headline">¶</a></h2>
<p>Here we build the RNAseq dataframe. Since TCGA gives us the genes identified by their Ensembl ID, we want to annotate these with their <code class="docutils literal notranslate"><span class="pre">gene</span> <span class="pre">symbol</span></code>. We do this by using biomart.</p>
<p>This section will take some time as we’re going to be downloading the annotation information from biomart.</p>
<p>Note you won’t have to do this everytime! It will just be the first time when you are creating this reference file.</p>
<p>The annotation file was generated by the following script from scibiomart:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">scibiomart</span> <span class="o">--</span><span class="n">m</span> <span class="n">ENSEMBL_MART_ENSEMBL</span> <span class="o">--</span><span class="n">d</span> <span class="n">hsapiens_gene_ensembl</span> <span class="o">--</span><span class="n">a</span> <span class="s2">&quot;ensembl_gene_id,entrezgene_id,external_gene_name,chromosome_name,start_position,end_position,strand&quot;</span> <span class="o">--</span><span class="n">o</span> <span class="n">tcga_</span>
</pre></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Build the RNAseq dataframe</span>
<span class="n">api</span><span class="o">.</span><span class="n">build_rna_df</span><span class="p">(</span><span class="n">download_dir</span><span class="p">)</span>

<span class="c1"># Now we need to access biomart to be able to mix between the two (i.e. one is annotated with gene symbol the other has ens id)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_rna_df</span><span class="p">()</span>

<span class="c1"># Use the pre stored annotation file, if you want to update this or add different attributes see scibiomart.</span>
<span class="c1"># api.set_gene_annotation_file(gene_annotation_file_path)</span>

<span class="c1"># Now we want to add our annotation data to our dataframe</span>
<span class="c1"># The first parameter is our dataframe, then the second parameter is whether we want to drop rows with null values</span>
<span class="c1"># We&#39;ll drop these rows since we only want data that has full metadata (i.e. go terms etc)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">add_gene_metadata_to_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Example-1:-Calculating-the-logfold-change-accross-2-conditions">
<h2>Example 1: Calculating the logfold change accross 2 conditions<a class="headerlink" href="#Example-1:-Calculating-the-logfold-change-accross-2-conditions" title="Permalink to this headline">¶</a></h2>
<p>Since the data from TCGA is just raw counts, we want to first normalise it. Usually I would use normalisations from DeSeq2, but for this tutorial we’ll just log transform counts + 1.</p>
<p>Here we’ll pretend to be interested in how female KIRC and male KIRC samples differ. To do this, we’ll first calculate the log fold change between the two conditions. Then to see if there are any outlier genes, we’ll have a look at a volcano plot of the data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">sciviso</span> <span class="kn">import</span> <span class="n">Boxplot</span><span class="p">,</span> <span class="n">Violinplot</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">filter_col</span> <span class="o">=</span> <span class="s1">&#39;ssm.consequence.0.transcript.gene.symbol&#39;</span>
<span class="n">column_returned</span> <span class="o">=</span> <span class="s2">&quot;case_id&quot;</span>
<span class="n">genes_of_interest</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;P53&#39;</span><span class="p">,</span> <span class="s1">&#39;SOX2&#39;</span><span class="p">,</span> <span class="s1">&#39;SOX1&#39;</span><span class="p">,</span> <span class="s1">&#39;SHH&#39;</span><span class="p">,</span> <span class="s1">&#39;HOXA1&#39;</span><span class="p">]</span>
<span class="n">gene_name</span> <span class="o">=</span> <span class="s1">&#39;external_gene_name&#39;</span>
<span class="n">ARL11_mutations</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_mutation_values_on_filter</span><span class="p">(</span><span class="n">column_returned</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;ARL11&#39;</span><span class="p">],</span> <span class="n">filter_col</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Samples with ARL11 mutations&quot;</span><span class="p">,</span> <span class="n">ARL11_mutations</span><span class="p">)</span>
<span class="n">transformed_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_values_from_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">ARL11_mutations</span><span class="p">)</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">gene_name</span> <span class="ow">and</span> <span class="s1">&#39;TCGA&#39;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
        <span class="n">transformed_df</span><span class="p">[</span><span class="s1">&#39;ARL11-MUT_&#39;</span> <span class="o">+</span> <span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

<span class="n">MUC4_mutations</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_mutation_values_on_filter</span><span class="p">(</span><span class="n">column_returned</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;MUC4&#39;</span><span class="p">],</span> <span class="n">filter_col</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Samples with MUC4 mutations&quot;</span><span class="p">,</span> <span class="n">MUC4_mutations</span><span class="p">)</span>

<span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_values_from_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">MUC4_mutations</span><span class="p">)</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">gene_name</span> <span class="ow">and</span> <span class="s1">&#39;TCGA&#39;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
        <span class="n">transformed_df</span><span class="p">[</span><span class="s1">&#39;MUC4-MUT_&#39;</span> <span class="o">+</span> <span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

<span class="c1"># Now lets calculate the changes between two conditions, let&#39;s go with KIRC vs KIRP</span>
<span class="n">transformed_df</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

<span class="n">boxplot</span> <span class="o">=</span> <span class="n">Boxplot</span><span class="p">(</span><span class="n">transformed_df</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">box_df</span> <span class="o">=</span> <span class="n">boxplot</span><span class="o">.</span><span class="n">format_data_for_boxplot</span><span class="p">(</span><span class="n">transformed_df</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;MUC4-MUT&#39;</span><span class="p">,</span> <span class="s1">&#39;ARL11-MUT&#39;</span><span class="p">],</span> <span class="n">gene_name</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;MUC4&quot;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">box_df</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<span class="n">boxplot</span> <span class="o">=</span> <span class="n">Violinplot</span><span class="p">(</span><span class="n">box_df</span><span class="p">,</span> <span class="s2">&quot;Conditions&quot;</span><span class="p">,</span> <span class="s2">&quot;Values&quot;</span><span class="p">,</span> <span class="s2">&quot;MUC4 MUT and ANK3 MUT&quot;</span><span class="p">,</span> <span class="n">add_dots</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">boxplot</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="c1"># Here normally we would run edgeR or deseq2 but we&#39;re just going to do a simple wilcoxin signed rank test between KIRP and KIRC for these conditions</span>


</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
No handles with labels found to put in legend.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Samples with ARL11 mutations [&#39;TCGA-IA-A83W&#39;]
Samples with MUC4 mutations [&#39;TCGA-P4-A5E8&#39;, &#39;TCGA-5P-A9JY&#39;, &#39;TCGA-UZ-A9PU&#39;, &#39;TCGA-IA-A83T&#39;, &#39;TCGA-SX-A7SM&#39;]
                                             Samples    Values Conditions
0  ARL11-MUT_TCGA-KIRP_PrimaryTumor_male_white_1_...  5.087463  ARL11-MUT
1  MUC4-MUT_TCGA-KIRP_PrimaryTumor_male_white_1_h...  6.918863   MUC4-MUT
2  MUC4-MUT_TCGA-KIRP_PrimaryTumor_male_white_1_h...  2.584963   MUC4-MUT
3  MUC4-MUT_TCGA-KIRP_SolidTissueNormal_male_whit...  5.754888   MUC4-MUT
4  MUC4-MUT_TCGA-KIRP_PrimaryTumor_male_blackoraf...  4.700440   MUC4-MUT
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook_12_2.png" src="../_images/examples_notebook_12_2.png" />
</div>
</div>
</div>
<div class="section" id="Example-2:-Looking-at-differences-across-genes-between-conditions">
<h2>Example 2: Looking at differences across genes between conditions<a class="headerlink" href="#Example-2:-Looking-at-differences-across-genes-between-conditions" title="Permalink to this headline">¶</a></h2>
<p>Here we have selected a group of genes and we’re interested in how specifc cases change across these genes. For this, we’ll use the log transformed data. Here we look at a box plot of each condition on the x axis and the y-values are the values for these conditions.</p>
<p>submitter_id project_id age_at_index gender race vital_status tumor_stage normal_samples tumor_samples case_files tumor_stage_num</p>
<p>SolidTissueNormal vs PrimaryTumor</p>
<p>For this, we’re just going to make two plots, a violin plot of the KIRP dataset for a set of genes and the KIRC we’re going to be looking at the normal tissue compared with the shape of the tumour tissue</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>


<span></span><span class="n">male_KIRP</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_files_with_meta</span><span class="p">({</span><span class="s1">&#39;gender&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;male&#39;</span><span class="p">],</span> <span class="s1">&#39;tumor_stage_num&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;project_id&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;TCGA-KIRP&#39;</span><span class="p">]})</span>
<span class="n">male_KIRC</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_files_with_meta</span><span class="p">({</span><span class="s1">&#39;gender&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;male&#39;</span><span class="p">],</span> <span class="s1">&#39;tumor_stage_num&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;project_id&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;TCGA-KIRC&#39;</span><span class="p">]})</span>

<span class="n">transformed_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="c1"># Here we&#39;re going to get a couple of groups of data, first the KIRP normal</span>
<span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_values_from_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">male_KIRP</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">column_name_includes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SolidTissueNormal&quot;</span><span class="p">])</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">gene_name</span> <span class="ow">and</span> <span class="s1">&#39;TCGA&#39;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
        <span class="n">transformed_df</span><span class="p">[</span><span class="s1">&#39;KIRP-NORMAL_&#39;</span> <span class="o">+</span> <span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_values_from_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">male_KIRP</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">column_name_includes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;PrimaryTumor&quot;</span><span class="p">])</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">gene_name</span> <span class="ow">and</span> <span class="s1">&#39;TCGA&#39;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
        <span class="n">transformed_df</span><span class="p">[</span><span class="s1">&#39;KIRP-TUMOUR_&#39;</span> <span class="o">+</span> <span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

<span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_values_from_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">male_KIRC</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">column_name_includes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SolidTissueNormal&quot;</span><span class="p">])</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">gene_name</span> <span class="ow">and</span> <span class="s1">&#39;TCGA&#39;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
        <span class="n">transformed_df</span><span class="p">[</span><span class="s1">&#39;KIRC-NORMAL_&#39;</span> <span class="o">+</span> <span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_values_from_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">male_KIRC</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">column_name_includes</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;PrimaryTumor&quot;</span><span class="p">])</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">gene_name</span> <span class="ow">and</span> <span class="s1">&#39;TCGA&#39;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
        <span class="n">transformed_df</span><span class="p">[</span><span class="s1">&#39;KIRC-TUMOUR_&#39;</span> <span class="o">+</span> <span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>


<span class="n">transformed_df</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

<span class="n">boxplot</span> <span class="o">=</span> <span class="n">Boxplot</span><span class="p">(</span><span class="n">transformed_df</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">box_df</span> <span class="o">=</span> <span class="n">boxplot</span><span class="o">.</span><span class="n">format_data_for_boxplot</span><span class="p">(</span><span class="n">transformed_df</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;KIRP-TUMOUR&#39;</span><span class="p">,</span> <span class="s1">&#39;KIRP-NORMAL&#39;</span><span class="p">,</span> <span class="s1">&#39;KIRC-NORMAL&#39;</span><span class="p">,</span> <span class="s1">&#39;KIRC-TUMOUR&#39;</span><span class="p">],</span> <span class="n">gene_name</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;ANK3&quot;</span><span class="p">])</span>
<span class="n">boxplot</span> <span class="o">=</span> <span class="n">Violinplot</span><span class="p">(</span><span class="n">box_df</span><span class="p">,</span> <span class="s2">&quot;Conditions&quot;</span><span class="p">,</span> <span class="s2">&quot;Values&quot;</span><span class="p">,</span> <span class="s2">&quot;Looking at KIRP v KIRC tumour and normal ANK3&quot;</span><span class="p">,</span> <span class="n">add_dots</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">boxplot</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
No handles with labels found to put in legend.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook_14_1.png" src="../_images/examples_notebook_14_1.png" />
</div>
</div>
</div>
<div class="section" id="Example-3:-Looking-at-genes-across-conditions">
<h2>Example 3: Looking at genes across conditions<a class="headerlink" href="#Example-3:-Looking-at-genes-across-conditions" title="Permalink to this headline">¶</a></h2>
<p>Here we may have a hypothesis related to how a specific gene changes across multiple conditions. For this, we may want to look at a box plot of the genes on the x-axis and the values of that gene from the different samples as the y axis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">sciviso</span> <span class="kn">import</span> <span class="n">Boxplot</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_rna_df</span><span class="p">()</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">add_gene_metadata_to_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

<span class="n">filter_col</span> <span class="o">=</span> <span class="s1">&#39;ssm.consequence.0.transcript.gene.symbol&#39;</span>
<span class="n">column_returned</span> <span class="o">=</span> <span class="s2">&quot;case_id&quot;</span>
<span class="n">genes_of_interest</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;SOX2&#39;</span><span class="p">,</span> <span class="s1">&#39;SOX1&#39;</span><span class="p">,</span> <span class="s1">&#39;SHH&#39;</span><span class="p">,</span> <span class="s1">&#39;HOXA1&#39;</span><span class="p">]</span>

<span class="c1"># Now we have our genes we want these to be our x axis</span>
<span class="n">female_stage1</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_files_with_meta</span><span class="p">({</span><span class="s1">&#39;gender&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;female&#39;</span><span class="p">],</span> <span class="s1">&#39;tumor_stage_num&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s1">&#39;project_id&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;TCGA-KIRP&#39;</span><span class="p">]})</span>
<span class="n">female_stage4</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_files_with_meta</span><span class="p">({</span><span class="s1">&#39;gender&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;female&#39;</span><span class="p">],</span> <span class="s1">&#39;tumor_stage_num&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;project_id&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;TCGA-KIRP&#39;</span><span class="p">]})</span>

<span class="c1"># Lets log2 + 1 the data</span>
<span class="n">transformed_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_values_from_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">female_stage4</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">filtered_df</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">gene_name</span> <span class="ow">and</span> <span class="s1">&#39;TCGA&#39;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
        <span class="n">transformed_df</span><span class="p">[</span><span class="s1">&#39;LATE_STAGE&#39;</span> <span class="o">+</span> <span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

<span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="p">,</span> <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">api</span><span class="o">.</span><span class="n">get_values_from_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">female_stage1</span><span class="p">)</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="n">gene_name</span> <span class="ow">and</span> <span class="s1">&#39;TCGA&#39;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">:</span>
        <span class="n">transformed_df</span><span class="p">[</span><span class="s1">&#39;EARLY_STAGE&#39;</span> <span class="o">+</span> <span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

<span class="n">transformed_df</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

<span class="c1"># Lets get the values for these cases</span>
<span class="c1"># Cool we have 56 and 19 that&#39;s enough for a statistical sig let&#39;s see if any of our genes are interesting</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes_of_interest</span><span class="p">:</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">boxplot</span> <span class="o">=</span> <span class="n">Boxplot</span><span class="p">(</span><span class="n">transformed_df</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="n">box_df</span> <span class="o">=</span> <span class="n">boxplot</span><span class="o">.</span><span class="n">format_data_for_boxplot</span><span class="p">(</span><span class="n">transformed_df</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;EARLY_STAGE&#39;</span><span class="p">,</span> <span class="s1">&#39;LATE_STAGE&#39;</span><span class="p">],</span> <span class="n">gene_name</span><span class="p">,</span> <span class="p">[</span><span class="n">gene</span><span class="p">])</span>
        <span class="n">boxplot</span> <span class="o">=</span> <span class="n">Boxplot</span><span class="p">(</span><span class="n">box_df</span><span class="p">,</span> <span class="s2">&quot;Conditions&quot;</span><span class="p">,</span> <span class="s2">&quot;Values&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">add_dots</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">boxplot</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(60684, 10)
p-value annotation legend:
ns: 5.00e-02 &lt; p &lt;= 1.00e+00
*: 1.00e-02 &lt; p &lt;= 5.00e-02
**: 1.00e-03 &lt; p &lt;= 1.00e-02
***: 1.00e-04 &lt; p &lt;= 1.00e-03
****: p &lt;= 1.00e-04

EARLY_STAGE v.s. LATE_STAGE: Mann-Whitney-Wilcoxon test two-sided with Bonferroni correction, P_val=9.052e-01 U_stat=2.550e+01
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook_16_1.png" src="../_images/examples_notebook_16_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
p-value annotation legend:
ns: 5.00e-02 &lt; p &lt;= 1.00e+00
*: 1.00e-02 &lt; p &lt;= 5.00e-02
**: 1.00e-03 &lt; p &lt;= 1.00e-02
***: 1.00e-04 &lt; p &lt;= 1.00e-03
****: p &lt;= 1.00e-04

EARLY_STAGE v.s. LATE_STAGE: Mann-Whitney-Wilcoxon test two-sided with Bonferroni correction, P_val=1.514e-01 U_stat=1.800e+01
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook_16_3.png" src="../_images/examples_notebook_16_3.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
p-value annotation legend:
ns: 5.00e-02 &lt; p &lt;= 1.00e+00
*: 1.00e-02 &lt; p &lt;= 5.00e-02
**: 1.00e-03 &lt; p &lt;= 1.00e-02
***: 1.00e-04 &lt; p &lt;= 1.00e-03
****: p &lt;= 1.00e-04

EARLY_STAGE v.s. LATE_STAGE: Mann-Whitney-Wilcoxon test two-sided with Bonferroni correction, P_val=2.159e-01 U_stat=3.800e+01
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook_16_5.png" src="../_images/examples_notebook_16_5.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
p-value annotation legend:
ns: 5.00e-02 &lt; p &lt;= 1.00e+00
*: 1.00e-02 &lt; p &lt;= 5.00e-02
**: 1.00e-03 &lt; p &lt;= 1.00e-02
***: 1.00e-04 &lt; p &lt;= 1.00e-03
****: p &lt;= 1.00e-04

EARLY_STAGE v.s. LATE_STAGE: Mann-Whitney-Wilcoxon test two-sided with Bonferroni correction, P_val=9.530e-01 U_stat=2.600e+01
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/examples_notebook_16_7.png" src="../_images/examples_notebook_16_7.png" />
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../faq.html" class="btn btn-neutral float-right" title="FAQ" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="cli.html" class="btn btn-neutral float-left" title="CLI" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Ariane Mora

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>